File: leetcode_solutions/2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        def linked_list_to_array(head):
            array = []
            current = head

            while current:
                array.append(current.val)
                current = current.next

            return array

        array = linked_list_to_array(head)
        i = 0
        j = len(array) - 1
        max_sum = float('-inf')

        while i < j:
            current_sum = array[i] + array[j]
            max_sum = max(max_sum, current_sum)
            i += 1
            j -= 1

        return max_sum
    
    
        

File: leetcode_solutions/2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.cs
public class Solution
{
    public int PairSum(ListNode head)
    {
        // Find the mid-point
        ListNode slow = head;
        ListNode fast = head.next;
        while (fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode head2 = slow.next;
        slow.next = null; // Cut the pointer/connection after mid-point
        
        // Reverse from mid point
        ListNode prev = null;
        ListNode curr = head2;
        while (curr != null)
        {
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }
        
        // Compute maximum twin sum 
        int res = int.MinValue;
        while(prev != null){
            res = Math.Max(res, head.val + prev.val);
            head = head.next;
            prev = prev.next;
        }
        
        return res;
    }
    

}
